include/deal.II/multigrid/mg_data_out.h

```diff
@@ -0,0 +1,528 @@
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_mg_data_out_h
+#define dealii_mg_data_out_h
+
+//#include <deal.II/multigrid/mg_base.h>
+#include <deal.II/base/mg_level_object.h>
+
+#include <deal.II/numerics/data_out.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/*
+
+- test subdomain id
+- test level
+- no dofhandler
+- with dofhandler
+- more than one component
+- parallel
+- matrix-free
+- trilinos
+
+ */
+
+
+/*!@addtogroup mg */
+/*@{*/
+
+
+template <int dim, typename DoFHandlerType = DoFHandler<dim>>
+class SingleLevelDataOut : public DataOut<dim, DoFHandlerType>
+{
+public:
+  SingleLevelDataOut()
+    : level_subdomain_id(-1)
+    , level(-1)
+  {}
+
+  SingleLevelDataOut(const unsigned int level_subdomain_id, unsigned int level)
+    : level_subdomain_id(level_subdomain_id)
+    , level(level)
+  {}
+
+  void
+  reinit(const unsigned int level_subdomain_id, unsigned int level)
+  {
+    this->level_subdomain_id = level_subdomain_id;
+    this->level              = level;
+  }
+
+
+  virtual typename DataOut<dim>::cell_iterator
+  first_cell()
+  {
+    typename DataOut<dim>::cell_iterator cell = this->dofs->begin(level);
+    while ((cell != this->dofs->end(level)) &&
+           (cell->level_subdomain_id() != level_subdomain_id))
+      ++cell;
+
+    if (cell == this->dofs->end(level))
+      return this->dofs->end();
+
+    return cell;
+  }
+
+  virtual typename DataOut<dim>::cell_iterator
+  next_cell(const typename DataOut<dim>::cell_iterator &old_cell)
+  {
+    if (old_cell != this->dofs->end(level))
+      {
+        typename DataOut<dim>::cell_iterator cell = old_cell;
+        ++cell;
+        while ((cell != this->dofs->end(level)) &&
+               (cell->level_subdomain_id() != level_subdomain_id))
+          ++cell;
+        if (cell == this->dofs->end(level))
+          return this->dofs->end();
+
+        return cell;
+      }
+    else
+      return this->dofs->end();
+  }
+
+  virtual typename DataOut<dim>::cell_iterator
+  first_locally_owned_cell()
+  {
+    return first_cell();
+  }
+
+
+  virtual typename DataOut<dim>::cell_iterator
+  next_locally_owned_cell(const typename DataOut<dim>::cell_iterator &old_cell)
+  {
+    return next_cell(old_cell);
+  }
+
+
+
+private:
+  unsigned int level_subdomain_id;
+  unsigned int level;
+};
+
+
+/**
+ * A specialized version of DataOut<dim> that supports output of cell
+ * and dof data on multigrid levels.
+ */
+template <int dim>
+class LevelDataOut : public DataOut<dim>
+{
+public:
+  /**
+   * Construct this output class. We will only output cells with the given
+   * @p subdomain_id and between level @p min_lvl and @p max_lvl. The default
+   * for @p max_lvl will output up to the finest existing level.
+   */
+  LevelDataOut(const unsigned int subdomain_id,
+               unsigned int       min_lvl = 0,
+               unsigned int       max_lvl = numbers::invalid_unsigned_int)
```

Comment:
Is it supposed to output the values on all levels between `min_lvl` and `max_lvl`? Also the values of cells refined that range? This would be too much, wouldn't it?

Reply:
Indeed -- I think that it would overwrite data from the coarser levels in the visualization. Why would you want to do that?